### PR TITLE
Support CommandPlugin

### PR DESCRIPTION
**Pull request by zunda-pixel**

### CODE REVIEW

File: Package.swift
Comment (allevato) [inline Package.swift:37]: I think we should remove the spaces from the _target_ names and subdirectory names. Names like `FormatPlugin` and `LintPlugin` would align well with the naming we've used elsewhere.

For the `name` argument of the `.plugin`, I'm less sure. I believe this surfaces in the Xcode UI? Is there anywhere else that this string is used (like a command line) where spaces would cause problems?

File: Package.swift
Comment (allevato) [inline Package.swift:117]: What's the behavior of `swift package format-source-code` when there are two plugins with this intent?

I wonder if we should use something like `.custom(verb: "lint-source-code")` here instead.

File: Plugins/Format Source Code/plugin.swift
Comment (allevato) [inline Plugins/Format Source Code/plugin.swift:30]: Here and throughout, can you use the `--long-form` of the flags instead, so it's easier for a reader to understand when they quickly scan it?

File: Plugins/Format Source Code/plugin.swift
Comment (allevato) [inline Plugins/Format Source Code/plugin.swift:35]: Here and elsewhere, let's avoid using new Swift 5.7 syntax for now; Package.swift still specifies a 5.6 tools version, and I'm not sure how it plays with CI (some other projects avoid it).

File: Plugins/Format Source Code/plugin.swift
Comment (allevato) [inline Plugins/Format Source Code/plugin.swift:19]: Instead of spawning `swift-format` once for each target, it would be more efficient to collect the directory paths in an array and then pass them all at once in the args array to just do a single invocation.

File: Plugins/Format Source Code/plugin.swift
Comment (allevato) [inline Plugins/Format Source Code/plugin.swift:60]: This code is mostly duplicated from the function above. Can you factor out the commonalities into a file-scoped function and just have the plugins call that?

I think a single function that takes the tool returned by `context.tool`, the arguments, and the list of directories to format would do the trick.

File: Plugins/Format Source Code/plugin.swift
Comment (allevato) [inline Plugins/Format Source Code/plugin.swift:68]: From the `XcodePluginContext` you can reach through `context.xcodeProject.directory`, which might be safer than assuming the working directory of the plugin will always be the project directory.

File: Plugins/Lint Source Code/plugin.swift
Comment (allevato) [inline Plugins/Lint Source Code/plugin.swift:1]: Same comments from the Format Source Code file apply here (factoring out the common implementations, making a single invocation to swift-format, accessing the Xcode project directory).

**Review by allevato — COMMENTED** [review]
Thanks for starting this! I haven't explored the plugin space much yet but this looks useful to have.

I'll caution you that the previous time I tried adding a plugin, we had to revert it because of CI and Windows-related failures: https://github.com/apple/swift-format/pull/424. But that was a _build tool_ plugin that was generated sources to use during swift-format's own build, so that might have been different. Either way, once this is in shape, we'll see if CI has any problems with it. 🤞🏻

File: Package.swift
Comment (zunda-pixel) [reply Package.swift]: I fixed with reference to [swift-argument-parser](https://github.com/apple/swift-argument-parser/blob/76259ebe8c477ecc4e938d0831aea4c25f9ee3ee/Package%40swift-5.6.swift)

[zunda-pixel/swift-format/pull/1](https://github.com/zunda-pixel/swift-format/pull/1)

Xcode  uses [Package.targets.plugin.name](https://github.com/zunda-pixel/swift-format/blob/41cf80954c9ab0bc8b3a562d89ad788534515649/Package.swift#L103) as UI

File: Package.swift
Comment (zunda-pixel) [reply Package.swift]: I fixed it. https://github.com/zunda-pixel/swift-format/commit/8bcb0bb94af287b080aa5d631360cc24737d9ba0

```shell
swift-format$ swift package plugin --list
‘format-source-code’ (plugin ‘Format Source Code’ in package ‘swift-format’)
‘generate-manual’ (plugin ‘Generate Manual’ in package ‘swift-argument-parser’)
‘lint-source-code’ (plugin ‘Lint Source Code’ in package ‘swift-format’)
```

File: Plugins/Format Source Code/plugin.swift
Comment (zunda-pixel) [reply Plugins/Format Source Code/plugin.swift]: I fix it. https://github.com/zunda-pixel/swift-format/commit/2fee7b9c1ad81955a054c02e6ca9884246b5d2b5

File: Plugins/Format Source Code/plugin.swift
Comment (zunda-pixel) [reply Plugins/Format Source Code/plugin.swift]: I fixed It.

File: Plugins/Format Source Code/plugin.swift
Comment (zunda-pixel) [reply Plugins/Format Source Code/plugin.swift]: I fixed it. https://github.com/zunda-pixel/swift-format/commit/60c66fc7a8404c4bbfd30506d38450042b292e5f

File: Plugins/Format Source Code/plugin.swift
Comment (zunda-pixel) [reply Plugins/Format Source Code/plugin.swift]: I don't know `--long-form`. what is this?

Comment (allevato) [reply Plugins/Format Source Code/plugin.swift]: Every flag that has a short form (like `-r`) has a longer form (`--recursive`). Using the longer ones will make the usage clearer. They're documented here: https://github.com/apple/swift-format#command-line-usage

File: Plugins/FormatPlugin/plugin.swift
Comment (zunda-pixel) [reply Plugins/FormatPlugin/plugin.swift]: I fixed it. https://github.com/apple/swift-format/pull/449/commits/28f9876c84f1252049a5bd87b83d41fdcb4438f1

**Review by allevato — APPROVED** [review]
Thank you! Waiting for CI results, which will be reported at https://github.com/apple/swift-syntax/pull/1086.

**Comment by allevato** [comment]
CI looks good. Can you squash the commits and re-push them so we can keep the history (even after merge) a bit cleaner?

**Comment by zunda-pixel** [comment]
I cleaned up git history.
I don't know git squash well, Is it ok?

**Review by allevato — APPROVED** [review]
This looks great; thank you again for the contribution!

**Pull request merged**
